drafts/str4d-overwinter-sighash/draft1.rst

```diff
@@ -0,0 +1,279 @@
+  increases. While a 1 MB block would normally take 2 seconds to verify with an average computer in 2015, a
+  1MB transaction with 5569 sigops may take 25 seconds to verify. This could be fixed by optimizing the digest
+  algorithm by introducing some reusable “midstate”, so the time complexity becomes O(n). [#quadratic]_
+
+* The algorithm does not involve the value being spent by the input. This is usually not a problem for online
+  network nodes as they could request for the specified transaction to acquire the output value. For an
+  offline transaction signing device ("cold wallet"), however, the lack of knowledge of input amount makes it
+  impossible to calculate the exact amount being spent and the transaction fee. To cope with this problem a
+  cold wallet must also acquire the full transaction being spent, which could be a big obstacle in the
+  implementation of a lightweight, air-gapped wallet. By including the input value of part of the transaction
+  digest, a cold wallet may safely sign a transaction by learning the value from an untrusted source. In the
+  case that a wrong value is provided and signed, the signature would be invalid and no funds would be lost.
+  [#offline-wallets]_
+
+Deploying the aforementioned fixes in the original script system is not a simple task.
+
+
+Specification
+=============
+
+A new transaction digest algorithm is defined, but only applicable from the Overwinter upgrade block height
+[#ZIP0000]_::
+  [TODO: Pick one] BLAKE2[b-256|s] of the serialization of:
+    1. nVersion of the transaction (4-byte little endian)
+    2. hashPrevouts (32-byte hash)
+    3. hashSequence (32-byte hash)
+    4. hashOutputs (32-byte hash)
+    5. hashJoinSplits (32-byte hash)
+    6. nLocktime of the transaction (8-byte little endian)
+    7. sighash type of the signature (4-byte little endian)
+    8. If we are serializing an input (ie. this is not a JoinSplit signature hash):
+       a. outpoint (32-byte hash + 4-byte little endian) 
+       b. scriptCode of the input (serialized as scripts inside CTxOuts) [TODO]
+       c. value of the output spent by this input (8-byte little endian)
+       d. nSequence of the input (4-byte little endian)
+
+The BLAKE2[b-256|s] personalization field will be set to::
+
+  ["ZcashSigHash"|"ZcSH"] || BRANCH_ID
+
+Semantics of the original sighash types remain unchanged, except the followings:
+
+#. The way of serialization is changed;
+
+#. All sighash types commit to the amount being spent by the signed input;
+
+#. ``SINGLE`` does not commit to the input index. When ``ANYONECANPAY`` is not set, the semantics are
+   unchanged since ``hashPrevouts`` and ``outpoint`` together implictly commit to the input index. When
+   ``SINGLE`` is used with ``ANYONECANPAY``, omission of the index commitment allows permutation of the
+   input-output pairs, as long as each pair is located at an equivalent index.
```

Comment:
(in reply to https://github.com/str4d/zips/commit/3d2923933d27648c8e2e6c2c0dfbfe511aded537#r26723093)

@daira I don't think there's any security issue. It just allows shuffling of the order of transparent input/output pairs, in the same way Sapling will allow shuffling of the order of input circuits and output circuits (not in pairs).